refactor(userspace/libsinsp)!: introducing plugin manager

jasondellaluce opened the pull request:
**What type of PR is this?**

/kind cleanup

/kind design

/kind feature

**Any specific area of the project related to this PR?**

/area libsinsp

**What this PR does / why we need it**:

This PR refactors the current libsinsp implementation of the plugin system by introducing the notion of "plugin manager". The manager is responsible of storing the plugins loaded at runtime by inspector instances and managing their state.

The manager is also used to implement optimized lookups, such as the source-compatibility check performed during field extraction (which used to be string-based and is now index-based). Right now, this should make a good performance optimization for plugin-based field extraction.

On top of that, the manager is also now the source of truth for the sources loaded at runtime, which can be useful in the future to support multiples source active at runtime.

**Which issue(s) this PR fixes**:

**Special notes for your reviewer**:

This PR brings few breaking changes in the `sinsp` class contract, which now exposes a new `get_plugin_manager()` method.

Moreover, the diff of the PR is big due to the git blame being moved across the splitted source files. The real diff size should be ~400 LOC.

**Does this PR introduce a user-facing change?**:

```release-note
refactor(userspace/libsinsp): introducing manager class for plugins
```

jasondellaluce commented:
/retest

jasondellaluce commented:
/retest

FedeDP commented:
I really like the idea! Code is much cleaner and simple this way, imo. 
That is the standard we have been using in libsinsp, so... :)

FedeDP commented on userspace/libsinsp/plugin_manager.cpp:
```diff
@@ -0,0 +1,101 @@
+{
+	auto it = m_plugins_id_source_index.find(plugin_id);
+	found = it != m_plugins_id_source_index.end();
+	return found ? it->second : 0;
```
Of course 0 is not a valid value there, i assume :)

jasondellaluce replied:
The semantic here is that if `bool& found` is set to false, the returned value is not significant. If you prefer, we can declare an "invalid return value" and return that instead, making it useless to have `bool& found` in the first place. Let me know what works best for you!

FedeDP replied:
Uh didn't notice that! Sorry, it's good for me ofc!

FedeDP reviewed (dismissed):
Left a minor comment; aside from that, this LGTM!
/approve

poiana commented:
LGTM label has been added.  <details>Git tree hash: 534e6713323c2b1c01344e3cff944a1817db3789</details>

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/313#pullrequestreview-967324229" title="Approved">FedeDP</a>*, *<a href="https://github.com/falcosecurity/libs/pull/313#" title="Author self-approved">jasondellaluce</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [FedeDP]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

Andreagit97 reviewed (commented):
Thank you @jasondellaluce! I think that this approach could significantly decrease the complexity around plugins :rocket: I left some minor inline comments

jasondellaluce commented on userspace/libsinsp/version.h:
```diff
@@ -0,0 +1,79 @@
+	{
+		if(this->m_version_major != requested.m_version_major)
+		{
+			// major numbers disagree
```
Since this implements a standard semantic versioning check, I don't see much value in printing out which part of the version check failed. If error handling is a concern, I would much prefer having more expressive errors at the level of the callers of this function.

Andreagit97 replied:
Yeah maybe you are right, since it is a standard check there is no need for further info. :+1:

poiana commented:
New changes are detected. LGTM label has been removed.

jasondellaluce commented:
Thanks you Andrea! I addressed your comments. On top of that, I realized that the getter methods of the plugin manager could be inlined to achieve better performance. I pushed a commit with all these changes, let me know what you think.

FedeDP commented on userspace/libsinsp/plugin_manager.cpp:
```diff
@@ -0,0 +1,67 @@
+		if(it->name() == plugin->name())
+		{
+			throw sinsp_exception(
+				"found multiple plugins with name " + it->name() + ". Aborting.");
```
```suggestion
				"found another plugin with name " + it->name() + ". Aborting.");
```

I'd rephrase it like this; i find it more correct. Wdyt?

poiana commented:
LGTM label has been added.  <details>Git tree hash: 9ab402058695a9ef48b0a773c665cc2c087d6de3</details>

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/313#pullrequestreview-969208759" title="Approved">FedeDP</a>*, *<a href="https://github.com/falcosecurity/libs/pull/313#" title="Author self-approved">jasondellaluce</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [FedeDP]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

Andreagit97 commented:
/approve

Andreagit97 reviewed (approved):
/approve

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/313#pullrequestreview-969292996" title="Approved">Andreagit97</a>*, *<a href="https://github.com/falcosecurity/libs/pull/313#pullrequestreview-969208759" title="Approved">FedeDP</a>*, *<a href="https://github.com/falcosecurity/libs/pull/313#" title="Author self-approved">jasondellaluce</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [Andreagit97,FedeDP]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/313#pullrequestreview-969292996" title="Approved">Andreagit97</a>*, *<a href="https://github.com/falcosecurity/libs/pull/313#pullrequestreview-969208759" title="Approved">FedeDP</a>*, *<a href="https://github.com/falcosecurity/libs/pull/313#" title="Author self-approved">jasondellaluce</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [Andreagit97,FedeDP]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->
